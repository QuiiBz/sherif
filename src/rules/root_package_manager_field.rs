use super::{Issue, IssueLevel};
use colored::Colorize;
use std::borrow::Cow;

#[derive(Debug)]
pub struct RootPackageManagerFieldIssue;

impl RootPackageManagerFieldIssue {
    pub fn new() -> Box<Self> {
        Box::new(Self {})
    }
}

impl Issue for RootPackageManagerFieldIssue {
    fn name(&self) -> &str {
        "root-package-manager-field"
    }

    fn level(&self) -> IssueLevel {
        IssueLevel::Error
    }

    fn message(&self) -> String {
        format!(
            r#"  │ {{
  {}   "{}": "..."   {}
  │ }}"#,
            "+".green(),
            "packageManager".white(),
            "← missing packageManager field.".green(),
        )
        .bright_black()
        .to_string()
    }

    fn why(&self) -> Cow<'static, str> {
        Cow::Borrowed("The root package.json should specify the package manager and version to use. Useful for tools like corepack.")
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test() {
        let issue = RootPackageManagerFieldIssue::new();

        assert_eq!(issue.name(), "root-package-manager-field");
        assert_eq!(issue.level(), IssueLevel::Error);

        colored::control::set_override(false);
        assert_eq!(
            issue.message(),
            "./package.json is missing `packageManager` field."
        );
    }
}
