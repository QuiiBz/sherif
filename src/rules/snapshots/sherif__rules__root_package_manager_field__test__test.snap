---
source: src/rules/root_package_manager_field.rs
expression: issue.message()
---
  │ {
  +   "packageManager": "..."   ← missing packageManager field.
  │ }
